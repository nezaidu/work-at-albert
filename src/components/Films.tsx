import moment from 'moment';
import React, {FC} from 'react';
import {View, TouchableOpacity} from 'react-native';
import {NavigationContext, useNavigation} from 'react-native-navigation-hooks/dist';
import {SmTitleText, SubtitleText, TitleText} from '../../ui';
import {Film as FilmType, SortingDirection} from '../generated/type';
import {
  FilmCard,
  FilmCountText,
  FilmsList,
  ToggleSortingIcon,
  ToggleSortingView,
} from './Films.styled';
import AscendingIcon from '../../assets/icons/Asc.svg';
import DescendingIcon from '../../assets/icons/Desc.svg';

interface FilmProps {
  film: FilmType;
  componentId: string;
}

const FilmItem: FC<FilmProps> = ({film, componentId}) => {
  const nav = useNavigation(componentId);

  return (
    <FilmCard
      onPress={() =>
        nav.push({
          component: {
            name: 'Film',
            passProps: {
              film,
            },
          },
        })
      }>
      <FilmCountText>{film.episodeID}</FilmCountText>
      <View>
        <TitleText>{film.title}</TitleText>
        <SmTitleText>{moment(film.releaseDate).local().year()}</SmTitleText>
      </View>
    </FilmCard>
  );
};

interface FilmsProps {
  data: FilmType[];
  componentId: string;
  sortingDirection: SortingDirection;
  toggleSorting: () => SortingDirection;
}

const Films: FC<FilmsProps> = ({data, sortingDirection, toggleSorting, componentId}) => {
  return (
    <FilmsList<FilmType>
      data={data}
      renderItem={({item}) => <FilmItem componentId={componentId} key={item.episodeID} film={item} />}
      ListHeaderComponent={() => (
        <View>
          <TouchableOpacity
            testID="toggle-sorting-button"
            onPress={toggleSorting}>
            {!!data.length && (
              <ToggleSortingView>
                <ToggleSortingIcon>
                  {sortingDirection === SortingDirection.Asc && (
                    <AscendingIcon />
                  )}
                  {sortingDirection === SortingDirection.Desc && (
                    <DescendingIcon />
                  )}
                </ToggleSortingIcon>
                <SubtitleText>
                  {moment(data[0].releaseDate).local().year()}...
                  {moment(data[data.length - 1].releaseDate)
                    .local()
                    .year()}
                </SubtitleText>
              </ToggleSortingView>
            )}
          </TouchableOpacity>
        </View>
      )}
    />
  );
};

export default Films;
