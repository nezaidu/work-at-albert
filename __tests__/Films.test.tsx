/**
 * @format
 */
import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import {fireEvent, render} from '@testing-library/react-native';
import initClient from '../src/core';
import filmsJSON from '../__mocks__/films.json';
import {Film, SortingDirection} from '../src/generated/type';
import Films from '../src/components/Films';
import {act} from 'react-test-renderer';

initClient();

const films: [Film] = filmsJSON as unknown as [Film];

const mockedType = require('../src/generated/type');

mockedType.useFilmsQuery = () => {
  return films;
};

it('render films', () => {
  const {toJSON} = render(
    <Films
      data={films}
      sortingDirection={SortingDirection.Asc}
      toggleSorting={jest.fn()}
    />,
  );

  const tree = toJSON();

  expect(tree).toMatchSnapshot();
});

it('should toggle sorting on button press', () => {
  const toggleSorting = jest.fn();
  const {getByTestId} = render(
    <Films
      data={films}
      sortingDirection={SortingDirection.Asc}
      toggleSorting={toggleSorting}
    />,
  );

  act(() => {
    fireEvent.press(getByTestId('toggle-sorting-button'));
  });

  expect(toggleSorting).toBeCalled();
});
