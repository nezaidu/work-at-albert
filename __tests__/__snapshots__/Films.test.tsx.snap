// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render films 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    Object {
      "data": Object {
        "allFilms": Object {
          "films": Array [
            Object {
              "episodeID": 4,
              "id": "ZmlsbXM6MQ==",
              "releaseDate": "1977-05-25",
              "title": "A New Hope",
            },
            Object {
              "episodeID": 5,
              "id": "ZmlsbXM6Mg==",
              "releaseDate": "1980-05-17",
              "title": "The Empire Strikes Back",
            },
            Object {
              "episodeID": 6,
              "id": "ZmlsbXM6Mw==",
              "releaseDate": "1983-05-25",
              "title": "Return of the Jedi",
            },
            Object {
              "episodeID": 1,
              "id": "ZmlsbXM6NA==",
              "releaseDate": "1999-05-19",
              "title": "The Phantom Menace",
            },
            Object {
              "episodeID": 2,
              "id": "ZmlsbXM6NQ==",
              "releaseDate": "2002-05-16",
              "title": "Attack of the Clones",
            },
            Object {
              "episodeID": 3,
              "id": "ZmlsbXM6Ng==",
              "releaseDate": "2005-05-19",
              "title": "Revenge of the Sith",
            },
          ],
          "totalCount": 6,
        },
      },
    }
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="toggle-sorting-button"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
